contracts/src/v0.8/dev/ocr2dr/OCR2DRRegistry.sol

```diff
@@ -0,0 +1,783 @@
+      // We want to ensure that we revert if gasAmount >  63//64*gas available
+      // as we do not want to provide them with less, however that check itself costs
+      // gas.  GAS_FOR_CALL_EXACT_CHECK ensures we have at least enough gas to be able
+      // to revert if gasAmount >  63//64*gas available.
+      if lt(g, GAS_FOR_CALL_EXACT_CHECK) {
+        revert(0, 0)
+      }
+      g := sub(g, GAS_FOR_CALL_EXACT_CHECK)
+      // if g - g//64 <= gasAmount, revert
+      // (we subtract g//64 because of EIP-150)
+      if iszero(gt(sub(g, div(g, 64)), gasAmount)) {
+        revert(0, 0)
+      }
+      // solidity calls check that a contract actually exists at the destination, so we do the same
+      if iszero(extcodesize(target)) {
+        revert(0, 0)
+      }
+      // call and return whether we succeeded. ignore return data
+      // call(gas,addr,value,argsOffset,argsLength,retOffset,retLength)
+      success := call(gasAmount, target, 0, add(data, 0x20), mload(data), 0, 0)
+    }
+    return success;
+  }
+
+  /**
+   * @inheritdoc OCR2DRRegistryInterface
+   */
+  function concludeBilling(
+    bytes32 requestId,
+    bytes calldata response,
+    bytes calldata err,
+    address transmitter,
+    address[MAX_NUM_ORACLES] memory, /* signers */
+    uint32 initialGas
+  ) external onlyAllowedDons nonReentrant returns (uint96) {
+    Commitment memory commitment = s_requestCommitments[requestId];
+    if (commitment.billing.client == address(0)) {
+      revert IncorrectCommitment();
+    }
+    delete s_requestCommitments[requestId];
+
+    bytes memory callback = abi.encodeWithSelector(
+      OCR2DRClientInterface.handleOracleFulfillment.selector,
+      requestId,
+      response,
+      err
+    );
+    // Call with explicitly the amount of callback gas requested
+    // Important to not let them exhaust the gas budget and avoid oracle payment.
+    // Do not allow any non-view/non-pure coordinator functions to be called
+    // during the consumers callback code via reentrancyLock.
+    // NOTE: that callWithExactGas will revert if we do not have sufficient gas
+    // to give the callee their requested amount.
+    s_config.reentrancyLock = true;
+    bool success = callWithExactGas(commitment.billing.gasLimit, commitment.billing.client, callback);
+    s_config.reentrancyLock = false;
+
+    // We want to charge users exactly for how much gas they use in their callback.
+    // The gasAfterPaymentCalculation is meant to cover these additional operations where we
+    // decrement the subscription balance and increment the oracles withdrawable balance.
+    uint96 payment = calculatePaymentAmount(
+      initialGas,
+      s_config.gasAfterPaymentCalculation,
+      commitment.donFee,
+      commitment.registryFee,
+      tx.gasprice
+    );
+    if (s_subscriptions[commitment.billing.subscriptionId].balance < payment) {
+      revert InsufficientBalance();
+    }
+    /**
+     * Oracle Payment *
```

Comment:
I can see a little mess of `Oracle` and `DON` terms. Can we use either DON or Oracle everywhere?

Reply:
Renamed it there. In some places it makes sense to keep Oracle (a single NOP on the DON).

I think we should rename `OCR2DROracle.sol` to make it more clear. What do you think?